cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/dataset/partitioned/DynamicPartitioningOutputCommitter.java

```diff
@@ -135,11 +139,28 @@ public void commitJob(JobContext context) throws IOException {
       ConfigurationUtil.getNamedConfigurations(this.taskContext.getConfiguration(),
                                                PartitionedFileSetArguments.OUTPUT_PARTITION_METADATA_PREFIX);
 
+    boolean allowAppend = partitionWriteOption == DynamicPartitioner.PartitionWriteOption.ALLOW_APPEND;
     // create all the necessary partitions
-    for (PartitionKey partitionKey : partitionsToAdd) {
-      PartitionOutput partitionOutput = outputDataset.getPartitionOutput(partitionKey);
-      partitionOutput.setMetadata(metadata);
-      partitionOutput.addPartition();
+    switch (partitionWriteOption) {
+      case ALLOW_APPEND:
+      case NEW_ONLY:
+        for (Map.Entry<String, PartitionKey> entry : partitionsToAdd.entrySet()) {
+          outputDataset.addPartition(entry.getValue(), entry.getKey(), metadata, true, allowAppend);
+        }
+        break;
+      case OVERWRITE:
+        for (Map.Entry<String, PartitionKey> entry : partitionsToAdd.entrySet()) {
+          if (outputDataset.getPartition(entry.getValue()) != null) {
+            // this allows reinstating the existing files if there's a rollback.
```

Comment:
ok that is good if there is already a test case for that. And the rollback happens in the reverse order? If you first drop, then recreate, then the rollback would have to undo the create first, and then undo the drop.